frame/staking/src/lib.rs

```diff
@@ -1270,6 +1453,100 @@ impl<T: Trait> Module<T> {
 
 	// MUTABLES (DANGEROUS)
 
+	fn do_payout_nominator(who: T::AccountId, era: EraIndex, validators: Vec<(T::AccountId, u32)>)
+		-> DispatchResult
+	{
+		let mut nominator_ledger = <Ledger<T>>::get(&who).ok_or_else(|| Error::<T>::NotController)?;
+		if nominator_ledger.next_reward > era {
+			return Err(Error::<T>::InvalidEraToReward.into());
+		}
+
+		nominator_ledger.next_reward = era + 1;
+		<Ledger<T>>::insert(&who, &nominator_ledger);
+
+		let mut reward = Perbill::zero();
+		let era_reward_points = <ErasRewardPoints<T>>::get(&era);
+
+		// Note: iterator take only `MAX_NOMINATIONS` to avoid querying more validator exposure
+		// than necessary. Anyway a nominator can't validate more than `MAX_NOMINATIONS`.
+		for (validator, nominator_index) in validators.into_iter().take(MAX_NOMINATIONS) {
+			let commission = Self::eras_validator_prefs(&era, &validator).commission;
+			let validator_exposure = <ErasStakersClipped<T>>::get(&era, &validator);
+
+			if let Some(nominator_exposure) = validator_exposure.others
+				.get(nominator_index as usize)
+			{
+				if nominator_exposure.who != nominator_ledger.stash {
+					continue;
+				}
+
+				let nominator_exposure_part = Perbill::from_rational_approximation(
+					nominator_exposure.value,
+					validator_exposure.total,
+				);
+				let validator_point = era_reward_points.individual.get(&validator)
+					.map(|points| *points)
+					.unwrap_or_else(|| Zero::zero());
+				let validator_point_part = Perbill::from_rational_approximation(
+					validator_point,
+					era_reward_points.total,
+				);
+				reward = reward.saturating_add(
+					validator_point_part
+						.saturating_mul(Perbill::one().saturating_sub(commission))
+						.saturating_mul(nominator_exposure_part)
+				);
+			}
+		}
+
+		// Note: this is zero if the era is not finished yet.
+		let era_payout = <ErasValidatorReward<T>>::get(&era);
+		if let Some(imbalance) = Self::make_payout(&nominator_ledger.stash, reward * era_payout) {
+			Self::deposit_event(RawEvent::Reward(who, imbalance.peek()));
+		}
+
+		Ok(())
+	}
+
+	fn do_payout_validator(who: T::AccountId, era: EraIndex) -> DispatchResult {
```

Comment:
This function should immediately fail if era > current era. A user can really screw themselves if we let them set "next_era" too high.

Really if we just track "last paid out era" and we put it into storage only after the payout, I think the user experience will be vastly improved. It allows someone to try and payout during a future era, the tx fails, and their "last era" tracker is untouched.

Reply:
> Really if we just track "last paid out era" and we put it into storage only after the payout, I think the user experience will be vastly improved. It allows someone to try and payout during a future era, the tx fails, and their "last era" tracker is untouched.

this doesn't change the issue, you point, if the user set a era index too high then he is screwed, user should claim their payout for `era < active_era` we can check that indeed.

Reply:
I now query EraRewardPayout as `Option<Balance>` if the era has no payout (era out of history or era not yet finished) then we just return InvalidEra error without fetching and upgrading ledger.last_reward